### PR TITLE
feat: Specify Node.js runtime version in serverless.yml

### PR DESCRIPTION
**Pull request by njm3754**
Specify Node.js runtime version in serverless.yml under provider. Throws an error for invalid Node.js version. 

AB#704

### CODE REVIEW

File: src/models/serverless.ts [hunk 1]
```diff
@@ -50,6 +50,7 @@ export interface ServerlessAzureProvider {
   hostingEnvironment?: ResourceConfig;
   virtualNetwork?: ResourceConfig;
   armTemplate?: ArmTemplateConfig;
+  runtime: string;
```
Comment (PIC123) [inline src/models/serverless.ts]: Does this need to be a required parameter?

Comment (njm3754) [reply src/models/serverless.ts]: Yes it is a required parameter based on the discussion

**Comment by njm3754** [comment]
Made some more changes. Now if the user does not specify runtime version the plugin throws an errors and just does not use the default value (10.14.1), Also added a new test case.

File: src/services/armService.ts [hunk 2]
```diff
@@ -190,4 +201,13 @@ export class ArmService extends BaseService {
...
     }
   }
+
+  private getNodeVersion(input: string): string{ 
+    const object = nodeVersions["nodejs"]; 
+    for(const nodeVersion of object){ 
+      if(nodeVersion["version"] == input || nodeVersion["type-latest"] == input) 
+        return nodeVersion["version"]; 
+    } 
+    return null;
```
Comment (tbarlow12) [inline src/services/armService.ts]: Throw an error here :)

File: src/services/armService.ts [hunk 3]
```diff
@@ -190,4 +201,13 @@ export class ArmService extends BaseService {
       });
     }
   }
+
+  private getNodeVersion(input: string): string{
```
Comment (tbarlow12) [inline src/services/armService.ts]: Be more explicit about the purpose of this function. Something like `validateNodeVersion`

File: src/services/armService.ts [hunk 4]
```diff
@@ -190,4 +201,13 @@ export class ArmService extends BaseService {
       });
     }
   }
+
+  private getNodeVersion(input: string): string{ 
+    const object = nodeVersions["nodejs"];
```
Comment (tbarlow12) [inline src/services/armService.ts]: Might be a better name than `object`

File: src/services/armService.ts [hunk 5]
```diff
@@ -190,4 +201,13 @@ export class ArmService extends BaseService {
       });
     }
   }
+
+  private getNodeVersion(input: string): string{ 
+    const object = nodeVersions["nodejs"]; 
+    for(const nodeVersion of object){
```
Comment (tbarlow12) [inline src/services/armService.ts]: Validation for major/minor/patch releases

File: src/armTemplates/resources/functionApp.ts [hunk 6]
```diff
@@ -117,4 +121,33 @@ export class FunctionAppResource implements ArmResourceTemplateGenerator {
...
   }
+
+  private getNodeVersion(runtime: string): string{ 
+    if(!runtime )
+      throw new Error("Node.js runtime version not specified in serverless.yml");
+
+    const nodeVersionObject = nodeVersions["nodejs"];
+
+    for(const nodeVersion of nodeVersionObject){ 
+      if(runtime.includes(nodeVersion["version"]) && semver.valid(nodeVersion["version"])){
```
Comment (tbarlow12) [inline src/armTemplates/resources/functionApp.ts]: Extract `nodeVersion["version"]` into a variable

File: src/armTemplates/resources/functionApp.ts [hunk 7]
```diff
@@ -117,4 +121,33 @@ export class FunctionAppResource implements ArmResourceTemplateGenerator {
       functionAppExtensionVersion: resourceConfig.extensionVersion,
     };
   }
+
+  private getNodeVersion(runtime: string): string{ 
+    if(!runtime )
+      throw new Error("Node.js runtime version not specified in serverless.yml");
+
+    const nodeVersionObject = nodeVersions["nodejs"];
```
Comment (tbarlow12) [inline src/armTemplates/resources/functionApp.ts]: The name `nodeVersionObject` is deceiving, I'd name it `nodeVersionList`

File: src/services/nodeVersion.json [hunk 8]
```diff
@@ -0,0 +1,63 @@
+{
```
Comment (tbarlow12) [inline src/services/nodeVersion.json]: Can you document where you got this file? Maybe in a comment in `functionApp.ts`?

File: src/armTemplates/resources/functionApp.ts [hunk 9]
```diff
@@ -117,4 +121,33 @@ export class FunctionAppResource implements ArmResourceTemplateGenerator {
...
+
+    for(const nodeVersion of nodeVersionObject){ 
+      if(runtime.includes(nodeVersion["version"]) && semver.valid(nodeVersion["version"])){
+        return nodeVersion["version"]; }
+    } 
+    
+    //Searches for the latest verison
+    if(runtime.includes(".x")){
+      let majorVersion =  runtime.replace(/[^0-9]/g,"");
```
Comment (tbarlow12) [inline src/armTemplates/resources/functionApp.ts]: Use `const` here

File: src/armTemplates/resources/functionApp.ts [hunk 10]
```diff
@@ -117,4 +121,33 @@ export class FunctionAppResource implements ArmResourceTemplateGenerator {
...
+      let nodeVersionsArray: string[] = [];
```
Comment (tbarlow12) [inline src/armTemplates/resources/functionApp.ts]: `const` here too

File: src/armTemplates/resources/functionApp.ts [hunk 11]
```diff
@@ -117,4 +121,33 @@ export class FunctionAppResource implements ArmResourceTemplateGenerator {
...
+    
+      for(const nodeVersion of nodeVersionObject){ 
+        let runtimeMajorVersion = nodeVersion["version"].split(".", 1);
```
Comment (tbarlow12) [inline src/armTemplates/resources/functionApp.ts]: const

File: src/armTemplates/resources/functionApp.ts (hunk 9 above)
Comment (tbarlow12) [reply src/armTemplates/resources/functionApp.ts]: Also, remove extra space

File: src/services/armService.test.ts [hunk 12]
```diff
@@ -93,7 +94,37 @@ describe("Arm Service", () => {
       await expect(service.createDeploymentFromType("not-found")).rejects.not.toBeNull();
     });
 
+    it("throws error when invalid nodejs version in defined", async () => {
+      sls.service.provider.runtime = "nodejs10.6.1"; 
+      await expect(service.createDeploymentFromType("premium")).rejects.toThrowError("Invalid Node.js version");
+    });
+
```
Comment (tbarlow12) [inline src/services/armService.test.ts]: Add tests for use case `major.minor.x`

**Pull request merged**